### PR TITLE
add code of conduct and community guidelines

### LINKED ISSUE #1: contribution guidelines and FAQ
i think the first step here is to document this fork and why it's there. lots of details are in https://github.com/wummel/linkchecker/issues/686, but it should be summarized in the documentation.

we should also formalize how we want to collaborate.

so - shopping list:

* [ ] document the fork and its reasons (summarize https://github.com/wummel/linkchecker/issues/686)
* [ ] document branches and protection - some of that done in #32, but needs details about specific branch support status
* [ ] add issue templates
* [x] document contribution guidelines - see #32
* [ ] adopt a code of conduct - see #32, not adopted yet
* [ ] add support section? FAQ, troubleshooting, proper bug reporting and so on... maybe based on [monkeysign's support section](http://monkeysign.readthedocs.io/en/2.x/support.html)?

### PR DESCRIPTION
**Pull request by anarcat**
Okay, here it goes... The objective here is to set clear guidelines on how to participate and contribute to this project. This is to ensure that what happened to linkchecker (the project was abandoned) doesn't happen again. It also ensures that we have a great project that is fun to participate in, without any problematic behavior that could ruin the work of our volunteers. It also tries to set examples on how to contribute on various levels - bug triage, patches and so on. It formalizes the review process we have experimented with.

More importantly, at this stage, it sets guidelines on how to share responsibility on the project. Right now, I have made a few people "members" here in a ad-hoc decision, but I have tried to formalize this process here as well.

Following the process defined here, at least one other member would need to approve this pull request before it is merged, at which point it would become policy. :)

This partly resolves #1. You'll notice there's a bunch of stuff missing, but there's only so much time and I want to get the ball rolling for the ownership, which seems to still be a critical SPOF with me in the middle. :)

### CODE REVIEW

**Comment by anarcat** [comment]
to see a rendered version, go at https://github.com/linkcheck/linkchecker/blob/a5dc2d861c73ade8412c8ad59033d60db73a99f0/doc/contributing.mdwn

maybe we'd need a table of contents?

what we really need is some online rendering of all this documentation, which should be merged with the website... so much to do!

**Comment by seamang** [comment]
I think this is generally a good idea though it doesn't get round the dependence problem - need more active people! Also, I think fundamental decisions about 'the constitution' should be an exception to the rule about just needing one person to approve; it's important that everyone agrees.

File: doc/contributing.mdwn [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+> Note: We have so far determined that writing an explicit enforcement
+> policy is not necessary, considering the available literature
+> already available online and the relatively small size of the
+> community. This may change in the future if the community grows
+> larger.
+
+# Patches
+
+Patches can be submitted through [pull requests][] on the
+[Github project][]. Some guidelines for patches:
```
Comment (mgedmin) [inline doc/contributing.mdwn]: I think the correct capitalization is GitHub.

File: doc/contributing.mdwn [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+-  try to reproduce the issue, if it is not reproducible, label it with
+   `help-wanted` and explain the steps taken to reproduce
+-  if information is missing, label it with `invalid` and request
+   specific information
+-  if the feature request is not within the scope of the project or
+   should be refused for other reasons, use the `wontfix` label and
+   close the issue
+-  mark feature requests with the `enhancement` label, bugs with
+   `bug`, duplicates with `duplicate` and so on...
```
Comment (mgedmin) [inline doc/contributing.mdwn]: It may be worth mentioning that only project members are allowed to label or close issues.

File: doc/contributing.mdwn [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+
+[issues]: https://github.com/linkcheck/linkchecker/issues
+
+# Membership
+
+There are three levels of membership in Github organizations, Owner,
+Member, or regular user. Anyone is welcome to contribute to the
+project within the guidelines outlined in this document, regardless of
+their status, and that includes regular users.
```
Comment (mgedmin) [inline doc/contributing.mdwn]: It is unclear to me if "regular users" are everyone with a GitHub account, or only those that have been invited into the linkcheck organization.

File: doc/development.mdwn [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+This is a technical document, if you are looking for ways to
+participate in the community, you should rather look into
+[contributing](contributing).
+
+Requirements
+------------
+On Mac OS X systems, using MacPorts, Fink or homebrew for software
+installation is recommended.
+
+- Install Python >= 2.7.2 from http://www.python.org/
```
Comment (mgedmin) [inline doc/development.mdwn]: Seems a rather low version. I think you need at least 2.7.6 for proper TLS support (e.g. SNI)?

File: doc/development.mdwn [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+- *On Windows only*, download and install the Microsoft
+  Visual C++ 2008 runtime from
+  http://www.microsoft.com/downloads/details.aspx?FamilyID=9b2da534-3e03-4391-8a4d-074b9f2bc1bf&displaylang=en
+
+- *Optional, used for Virus checking:*
+  ClamAv for Unix from http://www.clamav.net/lang/en/download/
+  or for Windows from http://www.sosdg.org/clamav-win32/
+
+- *Optional, for displaying country codes:*
+  Pygeoip from http://code.google.com/p/pygeoip/
```
Comment (mgedmin) [inline doc/development.mdwn]: The url says "This project has been moved over to github and will be deleted soon: https://github.com/appliedsec/pygeoip"

But anyway https://pypi.python.org/pypi/pygeoip exists, so maybe we should just add it to our install_requires and/or requirements.txt and drop the mention here?

File: doc/development.mdwn (hunk 4 above)
Comment (mgedmin) [inline doc/development.mdwn]: Also, can we make all the download URLs `https://`, provided they work?

File: doc/development.mdwn [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+
+Release process
+---------------
+
+1. make sure tests pass
+
+2. bump AppVersion in setup.py and commit changes
+
+3. build wheel file (setup.py bdist_wheel) and upload to PyPI
```
Comment (mgedmin) [inline doc/development.mdwn]: sdist too!

File: doc/development.mdwn [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+4. tag release (vX.Y.Z)
+
+5. push
+
+6. write release notes on github
+
+# Contributor Covenant Code of Conduct
```
Comment (mgedmin) [inline doc/development.mdwn]: Do we want this repeated twice in two files?

Comment (anarcat) [inline doc/development.mdwn]: definitely not, fixed. :)

File: doc/development.mdwn (hunk 6 above)
Comment (anarcat) [inline doc/development.mdwn]: in other projects, i keep source files on github because sdist files are not reproducible from the git tree directly, so they are different from github. so i keep only binary files on PyPI... but i'm fine with changing that - anyways, we don't have access to the PyPI archive (yet?) so this is mostly theoritical ... (#4)

fixed, anyways.

File: doc/development.mdwn (hunk 5 above)
Comment (anarcat) [inline doc/development.mdwn]: it seems you are reviewing the whole development.txt file - which i just renamed, i didn't actually change it that much... maybe we can split off those suggestions in a different PR? :)

File: doc/contributing.mdwn (hunk 3 above)
Comment (anarcat) [inline doc/contributing.mdwn]: tried to clarify.

File: doc/contributing.mdwn (hunk 2 above)
Comment (anarcat) [inline doc/contributing.mdwn]: tried to clarify.

File: doc/contributing.mdwn [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+  characters) line stating the a summary of the change, followed by a
+  blank line and then a description of the problem being solved and
+  its solution, or a reason for the change. Write more information,
+  not less, in the commit log.
+
+Patches should be reviewed by at least one other maintainer.
+
+Maintainers should not merge their own patches unless there is no
+response from other maintainers within a reasonable time frame (1-2
+days).
```
Comment (anarcat) [inline doc/contributing.mdwn]: agreed, I tried to rephrase.

File: doc/contributing.mdwn [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+
+[pull requests]: https://github.com/linkcheck/linkchecker/pulls
+[Github project]: https://github.com/linkcheck/linkchecker
+
+Some guidelines for patches:
+
+* A patch should be a minimal and accurate answer to exactly one
+  identified and agreed problem.
+* A patch must compile cleanly and pass project self-tests on at least
+  the principle target platform.
```
Comment (anarcat) [inline doc/contributing.mdwn]: actually, all supported platforms must pass (regardless of what those are) so i just changed that...

File: doc/contributing.mdwn [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+# Patches
+
+Patches can be submitted through [pull requests][] on the
+[Github project][]. Some guidelines for patches:
+
+[pull requests]: https://github.com/linkcheck/linkchecker/pulls
+[Github project]: https://github.com/linkcheck/linkchecker
+
+Some guidelines for patches:
```
Comment (anarcat) [inline doc/contributing.mdwn]: fixed

File: doc/contributing.mdwn (hunk 1 above)
Comment (anarcat) [inline doc/contributing.mdwn]: fixed everywhere in that document...

File: doc/development.mdwn (hunk 4 above)
Comment (anarcat) [inline doc/development.mdwn]: i did not actually change this - just renamed the file...

**Comment by seamang** [comment]
> A patch must compile cleanly and pass project self-tests on all target platforms.

Do we refuse pull requests from Unix users who haven't tested on Windows and vice versa? Or does this mean that our target platforms have to be limited to what Travis supports? I don't think a 'must' here is practical.

**Comment by anarcat** [comment]
okay, pushed a bunch more changes. @seamang i added an explicit rule about changing the constitution, that it requires all "adminstrators" (which i use instead of "owners" now, see 74487ae for the rationale) to have consensus.

i have also tried to address @mgedmin's excellent review. sorry for the mess about the duplicate stuff. also note that i didn't really change the development.txt file, i just renamed it to .mdwn which is why it looks like a new file, but it's just copy-paste, really. although it's true the release process is my own stuff, which is why i agree to change it here, it was changed in a previous (unreviewed!) push...

i agree we need more active people, but i think this will come with time... hopefully this document will help rather than be a nuisance. ;)

@seamang please do review the PR as well. ;)

**Comment by anarcat** [comment]
> Do we refuse pull requests from Unix users who haven't tested on Windows and vice versa? Or does this mean that our target platforms have to be limited to what Travis supports? I don't think a 'must' here is practical.

Well, we just need to clarify somewhere what our target platforms are. I don't think the statement should be interpreted as "we refuse code that wasn't tested on all platforms", but more as "we refuse code that fails to pass tests on all platforms", which is quite different. that we don't have a Windows test platform doesn't force (IMHO) UNIX users to install Windows 95 to test their shit. ;)

File: doc/development.mdwn (hunk 5 above)
Comment (mgedmin) [inline doc/development.mdwn]: Hey, free review for text I wouldn't have othervise reviewed ;)

**Comment by mgedmin** [comment]
Requiring every contributer to test on all supported platforms is not practical; that's why services like Travis CI (for Linux and Mac OS) and Appveyor exists.

We don't currently have Appveyor (or Mac OS targets for Travis CI) set up, but we should have, if we intend to support those platforms.  I can help with appveyor.yml, but I've never touched a Mac, so I hope somebody else steps up!

File: doc/contributing.mdwn [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+[issues]: https://github.com/linkcheck/linkchecker/issues
+
+Note that some of those operations are available only to project
+maintainers, see below for the different statuses.
+
+# Membership
+
+There are three levels of membership in the project, Administrator
+(also known as "Owner" in GitHub), Maintainer (also known as "Member")
+, or regular users (everyone with or without a GitHub account). Anyone
```
Comment (mgedmin) [inline doc/contributing.mdwn]: The newline in front of a comma seems weird to me.

I'm still tempted to nitpick about "membership in the project".  I think there are two levels of membership, but an important point is that you don't have to be a member to contribute.

Comment (anarcat) [inline doc/contributing.mdwn]: fixed the comma.

i welcome modifications on the wording of course. ;)

**Comment by anarcat** [comment]
> Requiring every contributer to test on all supported platforms is not practical; that's why services like Travis CI (for Linux and Mac OS) and Appveyor exist.

that's right.

> We don't currently have Appveyor (or Mac OS targets for Travis CI) set up, but we should have, if we intend to support those platforms. I can help with appveyor.yml, but I've never touched a Mac, so I hope somebody else steps up!

I thought travis had mac support? https://docs.travis-ci.com/user/multi-os/

the problem is really windows, and that's why we need Appveyor, AFAIK. i'm open to all of those. :)

**Comment by anarcat** [comment]
@seamang i think we only need your final approval before we go forward here.

File: doc/contributing.mdwn (hunk 2 above)
Comment (seamang) [inline doc/contributing.mdwn]: > A patch commit message must consist of a single short (less than 50  characters) line stating the a summary of the change, followed by a blank line and then a description of the problem being solved and its solution, or a reason for the change. Write more information, not less, in the commit log.

1. 'the a' -> a
 2. The 'commit log' is /doc/changelog.txt, in linkchecker. Maybe worth mentioning explicitly as it's not immediately obvious there is one. There isn't one in linkchecker-gui, needs creating.

Comment (anarcat) [inline doc/contributing.mdwn]: fixed the typo. i think we should keep changelog.txt to broader changes than just every commit log entry - those are separate things... changelog.txt is part of the release process, IMHO, which we need to improve, in doc/development.mdwn...

**Comment by anarcat** [comment]
awesome, thanks everyone for your reviews and feedback!

**Comment by anarcat** [comment]
oh, and i forgot to ask for another person to be added to the CoC contact list - see #33 for followup.

**Comment by mgedmin** [comment]
> I thought travis had mac support? https://docs.travis-ci.com/user/multi-os/

Yes, but we don't have it set up in our `.travis.yml` at the moment.  If you want an example of how other projects use Travis CI to build Mac OS X wheels for various Python versions, you can take a look at https://github.com/zopefoundation/zope.interface/blob/master/.travis.yml#L19-L33

**Comment by anarcat** [comment]
sounds like another PR waiting to be born. ;)

**Comment by mgedmin** [comment]
We could use a volunteer Mac maintainer.

I might be convinced to step up, if somebody wants to ship me a Macbook.

**Comment by mgedmin** [comment]
> I might be convinced to step up, if somebody wants to ship me a Macbook.

This sounds as if I'm serious.  To clarify: I'm not.  ;)

**Comment by anarcat** [comment]
maybe that could just be a differnt issue? :)
